core/src/datayoga_core/job.py

```diff
@@ -125,6 +125,49 @@ def transform(self, data: List[Dict[str, Any]], deepcopy: boolean = True) -> Job
 
         return result
 
+    async def async_transform(self, data: List[Dict[str, Any]], deepcopy: boolean = True) -> JobResult:
```

Comment:
can we have shared logic in `async_transform` and `transform` in a utility called from those methods rather than repeating ourselves? 95% of the logic is the same, will be tricky to maintain.

Reply:
one way suggested by chatgpt:
```
import asyncio

def transform(self, data: List[Dict[str, Any]], deepcopy: bool = True) -> JobResult:
    """
    Transforms data

    Args:
        data (List[Dict[str, Any]]): Data
        deepcopy: if True, performs a deepcopy before modifying records. otherwise, modifies in place. can affect performance.

    Returns:
        JobResult: Job result
    """
    if not self.initialized:
        logger.debug("job has not been initialized yet, initializing...")
        self.init()
    
    # Shared logic for both synchronous and asynchronous versions
    transformed_data = marshal.loads(marshal.dumps(data)) if deepcopy else data
    result = JobResult()
    
    # create an event loop for the synchronous version
    loop = asyncio.new_event_loop()

    try:
        # Call the async_transform method
        result = loop.run_until_complete(self.async_transform(transformed_data, deepcopy))
    finally:
        # close the event loop
        with suppress(NotImplementedError):
            # for pyodide. doesn't implement loop.close, ignore
            loop.close()

    return result

async def async_transform(self, data: List[Dict[str, Any]], deepcopy: bool = True) -> JobResult:
    """
    Transforms data (async version)

    Args:
        data (List[Dict[str, Any]]): Data
        deepcopy: if True, performs a deepcopy before modifying records. otherwise, modifies in place. can affect performance.

    Returns:
        JobResult: Job result
    """
    if not self.initialized:
        logger.debug("job has not been initialized yet, initializing...")
        self.init()

    # Shared logic for both synchronous and asynchronous versions
    transformed_data = marshal.loads(marshal.dumps(data)) if deepcopy else data
    result = JobResult()

    for step in self.steps:
        try:
            if len(transformed_data) == 0:
                # in case all records have been filtered, stop sending
                break
            processed, filtered, rejected = await step.block.run(transformed_data)
            result.filtered.extend(filtered)
            result.rejected.extend(rejected)
            transformed_data = [result.payload for result in processed]
        except ConnectionError as e:
            # connection errors are thrown back to the caller to handle
            raise e
        except Exception as e:
            # other exceptions are rejected
            logger.error(f"Error while transforming data: {e}")
            result.rejected.extend(
                [Result(Status.REJECTED, payload=row, message=f"{e}") for row in transformed_data])
            return result

    # the processed records are those that make it to the end
    result.processed = [Result(Status.SUCCESS, payload=row) for row in transformed_data]

    return result

```

Not sure if we want to create another event loop in the synced version, but an idea to consider.

Reply:
This one doesn't create another event loop:
```
import asyncio

def transform(self, data: List[Dict[str, Any]], deepcopy: bool = True) -> JobResult:
    """
    Transforms data

    Args:
        data (List[Dict[str, Any]]): Data
        deepcopy: if True, performs a deepcopy before modifying records. otherwise, modifies in place. can affect performance.

    Returns:
        JobResult: Job result
    """
    if not self.initialized:
        logger.debug("job has not been initialized yet, initializing...")
        self.init()

    # Shared logic for both synchronous and asynchronous versions
    transformed_data = marshal.loads(marshal.dumps(data)) if deepcopy else data
    result = asyncio.run(self.async_transform(transformed_data, deepcopy))
    
    return result

async def async_transform(self, data: List[Dict[str, Any]], deepcopy: bool = True) -> JobResult:
    """
    Transforms data (async version)

    Args:
        data (List[Dict[str, Any]]): Data
        deepcopy: if True, performs a deepcopy before modifying records. otherwise, modifies in place. can affect performance.

    Returns:
        JobResult: Job result
    """
    if not self.initialized:
        logger.debug("job has not been initialized yet, initializing...")
        self.init()

    # Shared logic for both synchronous and asynchronous versions
    transformed_data = marshal.loads(marshal.dumps(data)) if deepcopy else data
    result = JobResult()

    for step in self.steps:
        try:
            if len(transformed_data) == 0:
                # in case all records have been filtered, stop sending
                break
            processed, filtered, rejected = await step.block.run(transformed_data)
            result.filtered.extend(filtered)
            result.rejected.extend(rejected)
            transformed_data = [result.payload for result in processed]
        except ConnectionError as e:
            # connection errors are thrown back to the caller to handle
            raise e
        except Exception as e:
            # other exceptions are rejected
            logger.error(f"Error while transforming data: {e}")
            result.rejected.extend(
                [Result(Status.REJECTED, payload=row, message=f"{e}") for row in transformed_data])
            return result

    # the processed records are those that make it to the end
    result.processed = [Result(Status.SUCCESS, payload=row) for row in transformed_data]

    return result
```

See if this can fit or suggest something else to refactor.